Convert patchviasocket to python (removes perl dependency for KVM agent)

sverrirab opened the pull request:
As requested here: https://github.com/apache/cloudstack/pull/1495

No scripts are using perl so that install requirement can be removed.
The new scripts are using standard python packages only.
Includes extensive unit test.
Note: perl-modules requirement is missing (fixed in mentioned PR) so do not merge that onto master.

wido commented on scripts/vm/hypervisor/kvm/patchviasocket.py:
```diff
@@ -0,0 +1,80 @@
+        s.connect(sock_file)
+        s.sendall(msg)
+        s.close()
+    except IOError as e:
```
Is IOError the only exception which can occur here?

Wouldn't it be safer to but a 'return 1' in a 'finally' statement and otherwise return 0?

sverrirab replied:
IOError is the parent of socket.error so this will catch all socket related errors.
The code as it stands returns 1 on any error and 0 on success - so there is no benefit of adding a finally statement there that I can see.

wido commented:
Thanks for this! In the future I'd like to see this done through the Qemu Guest Agent, but that is still waiting for something, see: #985

jburwell commented on scripts/vm/hypervisor/kvm/patchviasocket.py:
```diff
@@ -0,0 +1,80 @@
+        if os.path.isfile(key_file):
+            with open(key_file, "r") as f:
+                return f.read()
+    except IOError:
```
Would it be helpful to print/log the details of the ``IOError`` for debugging purposes?  For example, understanding whether the file was not found or the user lacked sufficient permissions.

jburwell commented on scripts/vm/hypervisor/kvm/patchviasocket.py:
```diff
@@ -0,0 +1,80 @@
+    pub_key = read_pub_key(key_file)
+
+    if not pub_key:
+        print("ERROR: ssh public key not found on host at {0}".format(key_file))
```
Why throw this error out of ``read_pub_key`` and consolidate the error handling/return in the exception handler on line 64?

jburwell commented on scripts/vm/hypervisor/kvm/test_patchviasocket.py:
```diff
@@ -0,0 +1,142 @@
+    def file(self):
+        return self._file
+
+    def wait_until_ready(self):
```
Should this be bounded by a maximum wait time?  For example, if the ulimits are exceeded, this function could turn into an infinite loop.

jburwell commented on scripts/vm/hypervisor/kvm/test_patchviasocket.py:
```diff
@@ -0,0 +1,142 @@
+    def __init__(self):
+        super(SocketThread, self).__init__()
+        self._data = ""
+        self._file = tempfile.mktemp(".sck")
```
According to the [Python API document for ``tempfile.mktemp``](https://docs.python.org/2/library/tempfile.html), this method has been deprecated since Python 2.3.  ``tempfile.mkstemp`` should be used instead.

jburwell commented on scripts/vm/hypervisor/kvm/test_patchviasocket.py:
```diff
@@ -0,0 +1,142 @@
+
+
+def write_key_file():
+    tmpfile = tempfile.mktemp(".sck")
```
According to the [Python API document for ``tempfile.mktemp``](https://docs.python.org/2/library/tempfile.html), this method has been deprecated since Python 2.3.  ``tempfile.mkstemp`` should be used instead.

jburwell commented on scripts/vm/hypervisor/kvm/test_patchviasocket.py:
```diff
@@ -0,0 +1,142 @@
+        except socket.timeout:
+            pass
+        s.close()
+        os.remove(self._file)
```
Line 73-74 should be in a ``finally`` block to ensure the resources are properly closed when an exception occurs.

DaanHoogland commented:
I hadn't looked into the code yet so for what it's worth and pending response to the comments:
CI RESULTS
[1533.results.internal_lb.txt](https://github.com/apache/cloudstack/files/253034/1533.results.internal_lb.txt)
[1533.results.network.txt](https://github.com/apache/cloudstack/files/253035/1533.results.network.txt)
[1533.results.vpc_routers.txt](https://github.com/apache/cloudstack/files/253037/1533.results.vpc_routers.txt)
[1533.results.vpc_vpn.txt](https://github.com/apache/cloudstack/files/253036/1533.results.vpc_vpn.txt)

I didn't bother with the known ping problem in test_02_redundant_VPC_default_routes

swill commented:
@DaanHoogland can you post a 'summary' of the results when you post this.  I don't want to have to download each of these files to know that they ran successfully.  Can you just say something like "CI Passed" or something like that?  "Here are the results" without telling me if it is a success for failure requires me to download and review every file.  I understand from your "dont bother with" comment that everything passed except that, but if you can just let me know that status, that will help me.  Thanks for testing this.  👍

swill commented:
I am missing some code review on this one, but otherwise this is looking good so far.

DaanHoogland commented:
Sorry Will, Yes all passed, some after retest.

wido commented:
Can @sverrirab look at the comments of @jburwell ?

sverrirab commented:
@wido I just pushed an update that addresses the comments from @jburwell 

tested this locally as well of course

wido commented:
Thanks! code-wise this looks good to me now

swill commented:
@sverrirab thank you.  Would you mind adding the revert of #1495 into this commit in order to clean up unnecessary packages once this PR is accepted?  I think we are missing one code review on this one and it is pretty much ready to go.  Thanks...

sverrirab commented:
Rebased the PR to latest master and reverted the relevant commit (64b72a5c5a410f41bd869cc9d40807d888e05055.).  I think we should be good to go @swill ?

swill commented:
### CI RESULTS

```
Tests Run: 85
  Skipped: 0
   Failed: 0
   Errors: 0
 Duration: 8h 55m 29s
```



**Associated Uploads**

**`/tmp/MarvinLogs/DeployDataCenter__May_21_2016_01_21_56_QYQ9XC:`**
* [dc_entries.obj](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1533/tmp/MarvinLogs/DeployDataCenter__May_21_2016_01_21_56_QYQ9XC/dc_entries.obj)
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1533/tmp/MarvinLogs/DeployDataCenter__May_21_2016_01_21_56_QYQ9XC/failed_plus_exceptions.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1533/tmp/MarvinLogs/DeployDataCenter__May_21_2016_01_21_56_QYQ9XC/runinfo.txt)

**`/tmp/MarvinLogs/test_network_T2P1UQ:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1533/tmp/MarvinLogs/test_network_T2P1UQ/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1533/tmp/MarvinLogs/test_network_T2P1UQ/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1533/tmp/MarvinLogs/test_network_T2P1UQ/runinfo.txt)

**`/tmp/MarvinLogs/test_vpc_routers_U3WSE5:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1533/tmp/MarvinLogs/test_vpc_routers_U3WSE5/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1533/tmp/MarvinLogs/test_vpc_routers_U3WSE5/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1533/tmp/MarvinLogs/test_vpc_routers_U3WSE5/runinfo.txt)


Uploads will be available until `2016-07-22 02:00:00 +0200 CEST`

*Comment created by [`upr comment`](https://github.com/cloudops/upr).*

swill commented:
This is coming back clean.  Would you mind trying to rebase and re-push or close and reopen to see if we can kick off travis one more time.  Thanks...

sverrirab commented:
everything looking good now it seems - time to merge @swill ?

swill commented:
Yes, ready now.  Thank you.  :)
